src/stdio_mgr/stdio_mgr.py

```diff
@@ -27,12 +27,58 @@
 """
 
 import sys
-from contextlib import contextmanager
-from io import BufferedReader, BytesIO, StringIO, TextIOBase, TextIOWrapper
+from contextlib import contextmanager, ExitStack, suppress
+from io import BufferedRandom, BufferedReader, BytesIO, TextIOBase, TextIOWrapper
 
 import attr
 
 
+class _PersistedBytesIO(BytesIO):
+    """Class to persist the stream after close.
+
+    A copy of the bytes value is available at :attr:`_closed_buf` after
+    the :meth:`~BytesIO.close`.
+    """
+
+    def close(self):
+        self._closed_buf = self.getvalue()
```

Comment:
Worth noting, that this buffer doesnt get deleted at all, so it could be a memory problem if the value was large.

I did consider a few ways to remove it -- the most obvious is to `del` it after it has been read once.

However it will be solved later at https://github.com/bskinn/stdio-mgr/issues/3 , as that would `del` the objects which refer to this buffer and it would then disappear.

Or a nice trick to explore is for the RandomTextIO objects to know when they are in context, and change their `.close` behaviour accordingly so that they are 'safe' closes only within the context manager, and also so the value persists after close only within the context.  i.e. when someone disables auto-close, the extra features are removed when the context is exited, and the user is now in manual mode.

Reply:
> Or a nice trick to explore is for the RandomTextIO objects to know when they are in context, and change their .close behaviour accordingly so that they are 'safe' closes only within the context manager, and also so the value persists after close only within the context. i.e. when someone disables auto-close, the extra features are removed when the context is exited, and the user is now in manual mode.

Might be as simple as overriding `__enter__` to set, and `__exit__` to clear, an internal flag that can be inspected within `close()` to determine the proper behavior?

Reply:
Im still unsure of what the 'proper' behaviour should be.  Having `getvalue` always work, including outside the context, could be very valuable in getting stdio-mgr adopted.

I did a sample of using stdio-mgr with the repo at https://github.com/bskinn/stdio-mgr/issues/3#issuecomment-515734883 , and the diff was huge, but it was removing horrible mocking with simple stdio-mgr -- but if I needed to also move asserts into the context, it becomes larger and harder to spot check.
(That also needed binary `init_text` so it cant adopt stdio-mgr quite yet)

.> Might be as simple as overriding __enter__ to set, and __exit__ to clear, an internal flag that can be inspected within close()

This could mean the separate 'safe' versions of the classes are not necessary.  I was toying with that idea before creating those safe classes, but went with those safe classes so that the caller can see that `close=False` gives them a more raw wrapper.  But it does feel like there is more value now in having 'safe-close'/other-side-effects mode be dynamic.

Also since you mentioned `__enter__`.  We should have some tests and documented behaviour for using stdio-mgr without `with`.  i.e. on Py37
```py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot unpack non-iterable _GeneratorContextManager object
>>> m = stdio_mgr(close=False)
>>> m
<contextlib._GeneratorContextManager object at 0x7f7d6b8d3be0>
>>> dir(m)
['__abstractmethods__', '__call__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_recreate_cm', 'args', 'func', 'gen', 'kwds']
>>> list(m.gen)
[(TeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8'), <_io.TextIOWrapper encoding='utf-8'>, <_io.TextIOWrapper encoding='utf-8'>)]
>>> m = stdio_mgr(close=False)
>>> (i, o, e) = list(m.gen)[0]
>>> i
TeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8')
>>> o
<_io.TextIOWrapper encoding='utf-8'>
>>> e
<_io.TextIOWrapper encoding='utf-8'>
>>> o.buffer
<_io.BufferedRandom>
>>> o.buffer.raw
<stdio_mgr.stdio_mgr._PersistedBytesIO object at 0x7f7d6b985200>
```
If we did change stdio_mgr to a class, the resulting class would be like a text fifo between i and o?
How can it be useful?  They can access it anyway using the above loop-hole.
This might also help when trying to come up with a class name for it, if the function is to be replaced with a class.
Or can we close the loop-hole, unless moving to a class.

Reply:
> Worth noting, that this buffer doesnt get deleted at all, so it could be a memory problem if the value was large.

As you noted elsewhere, as the anticipated major `stdio-mgr` use case is testing, memory efficiency is not a big concern ATM. Optimizing memory footprint may well be a giant YAGNI. Let's not worry about this for now.

-----

> Im still unsure of what the 'proper' behaviour should be. Having getvalue always work, including outside the context, could be very valuable in getting stdio-mgr adopted.

Yeah, I really like the prospect of `getvalue()` always working.  I think that should be a core feature/selling point. If there aren't tests confirming that it's available in all cases, can you add?

----

> (That also needed binary `init_text` so it cant adopt stdio-mgr quite yet)

Supporting `init_text` as `bytes` added as #36.

----

> This could mean the separate 'safe' versions of the classes are not necessary. I was toying with that idea before creating those safe classes, but went with those safe classes so that the caller can see that close=False gives them a more raw wrapper. But it does feel like there is more value now in having 'safe-close'/other-side-effects mode be dynamic.

Created as #37.

----

> Also since you mentioned `__enter__`. We should have some tests and documented behaviour for using stdio-mgr without `with`. i.e. on Py37

>
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot unpack non-iterable _GeneratorContextManager object
>>> m = stdio_mgr(close=False)
>>> m
<contextlib._GeneratorContextManager object at 0x7f7d6b8d3be0>
>>> dir(m)
['__abstractmethods__', '__call__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_recreate_cm', 'args', 'func', 'gen', 'kwds']
>>> list(m.gen)
[(TeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8'), <_io.TextIOWrapper encoding='utf-8'>, <_io.TextIOWrapper encoding='utf-8'>)]
>>> m = stdio_mgr(close=False)
>>> (i, o, e) = list(m.gen)[0]
>>> i
TeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8')
>>> o
<_io.TextIOWrapper encoding='utf-8'>
>>> e
<_io.TextIOWrapper encoding='utf-8'>
>>> o.buffer
<_io.BufferedRandom>
>>> o.buffer.raw
<stdio_mgr.stdio_mgr._PersistedBytesIO object at 0x7f7d6b985200>
```
> If we did change stdio_mgr to a class, the resulting class would be like a text fifo between i and o?
> How can it be useful? They can access it anyway using the above loop-hole.
> This might also help when trying to come up with a class name for it, if the function is to be replaced with a class.
> Or can we close the loop-hole, unless moving to a class.

Created #38 for this.